src/main/java/com/google/cloud/teleport/templates/PubSubToCloudStorageCompressed.java

```diff
@@ -0,0 +1,222 @@
+    @Required
+    ValueProvider<String> getOutputFilenamePrefix();
+    void setOutputFilenamePrefix(ValueProvider<String> value);
+
+    @Description("The suffix of the files to write.")
+    @Default.String("")
+    ValueProvider<String> getOutputFilenameSuffix();
+    void setOutputFilenameSuffix(ValueProvider<String> value);
+
+    @Description("The shard template of the output file. Specified as repeating sequences "
+        + "of the letters 'S' or 'N' (example: SSS-NNN). These are replaced with the "
+        + "shard number, or number of shards respectively")
+    @Default.String("W-P-SS-of-NN")
+    ValueProvider<String> getOutputShardTemplate();
+    void setOutputShardTemplate(ValueProvider<String> value);
+
+    @Description("The maximum number of output shards produced when writing.")
+    @Default.Integer(1)
+    Integer getNumShards();
+    void setNumShards(Integer value);
+
+    @Description("The window duration in which data will be written. Defaults to 5m. "
+        + "Allowed formats are: "
+        + "Ns (for seconds, example: 5s), "
+        + "Nm (for minutes, example: 12m), "
+        + "Nh (for hours, example: 2h).")
+    @Default.String("5m")
+    String getWindowDuration();
+    void setWindowDuration(String value);
+
+    @Description("The compression algorithm to use on the matched files.")
+    @Required
+    Compression getCompression();
+
+    void setCompression(Compression value);
+  }
+
+  /**
+   * Main entry point for executing the pipeline.
+   * @param args  The command-line arguments to the pipeline.
+   */
+  public static void main(String[] args) {
+
+    Options options = PipelineOptionsFactory
+        .fromArgs(args)
+        .withValidation()
+        .as(Options.class);
+
+    options.setStreaming(true);
+
+    run(options);
+  }
+
+  /**
+   * Runs the pipeline with the supplied options.
+   *
+   * @param options The execution parameters to the pipeline.
+   * @return  The result of the pipeline execution.
+   */
+  public static PipelineResult run(Options options) {
+    // Create the pipeline
+    Pipeline pipeline = Pipeline.create(options);
+
+    /*
+     * Steps:
+     *   1) Read string messages from PubSub
+     *   2) Window the messages into minute intervals specified by the executor.
+     *   3) Output the windowed files to GCS
+     */
+    pipeline
+        .apply("ReadPubSubSubscription", PubsubIO.readStrings().fromSubscription(options.getInputSubscription()))
+        .apply(
+            options.getWindowDuration() + " Window",
+            Window.into(FixedWindows.of(DurationUtils.parseDuration(options.getWindowDuration()))))
+
+        // Apply windowed file writes. Use a NestedValueProvider because the filename
+        // policy requires a resourceId generated from the input value at runtime.
+        .apply(
+            "Write File(s)",
+            TextIO.write()
+                .withWindowedWrites()
+                .withCompression(options.getCompression())
```

Comment:
Since compression does not have a ValueProvider, the pipeline would use the static value available at _compile time_. This would not be something that the user can change at _runtime_. i.e For example, if the template were compiled with a compression value of GZIP, then the template would always use GZIP as the compression regardless of what is specified at runtime.

Reply:
Oooh I see now. Thank you I will update compression to use the a provider.